src/resolve/resolved-dns-rr.c

```diff
@@ -826,6 +827,115 @@ static char *format_txt(DnsTxtItem *first) {
         return s;
 }
 
+static char *format_svc_param_value(DnsSvcParam *i) {
+        char *value = NULL;
+        int r;
+
+        switch (i->key) {
+        case DNS_SVC_PARAM_KEY_ALPN: {
+                size_t sz = 0, offset = 0;
+                _cleanup_strv_free_ char **values_strv = NULL;
+                while (offset < i->length) {
+                        char *alpn;
+                        sz = (uint8_t) i->value[offset++];
+                        alpn = strndup((char *)&i->value[offset], sz);
+                        if (!alpn)
+                                return NULL;
+                        strv_push(&values_strv, alpn);
+                        offset += sz;
```

Comment:
there probably should be some validation here that the string doesn't carry control characters, or maybe even is valid UTF8? or alternatively, there should be escaping in place (via cescape()?) to deal with weird bytes

Reply:
Sadly this field is explicitly _not_ valid UTF-8, there is actually a list of known values in iana: https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids, several reserved values have non utf-8 byte sequences. Seems like the useful values are human-readable strings though — so yeah I think c-style escape here will do.

I'll fix this.